### ISSUE TITLE
SelectOneMenu: ARIA compliance

### ISSUE THREAD

**Issue by Bukama** [primefaces/primefaces#9279]
### Description

[Primeface's SelectOneMenu](https://www.primefaces.org/showcase/ui/input/oneMenu.xhtml?) simulates a combobox using a `div` with `role="combobox"` and a hidden `input` element (in showcase it has the id `j_idt343:option_focus` ) for the current selected value.
To this hidden input element has the role `text` is assigned, as colored in the screenshot.

When using a screenreader on a page with such an combobox it reads out

> text input with auto-completion empty

This might be a bit confused so let's get the unproblematic parts ot it out of the way first:
* "auto-completion" comes from `aria-autocomplete="list"` which is absolutely correct set, see [Moz Dev, ARIA autocomplete values](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-autocomplete#values)
* "empty" comes from a not yet selected value and is updated when selecting one

So this leaves us with the first value "text input", which comes from the `role` attribute of the hidden input element and indicates that on the page there is an text input element instead of an combobox.

According to the accessibilty test team at my work this is problematic for users, who are visible impaired.

![primefaces_role_selectOneMenu](https://user-images.githubusercontent.com/4354359/194518757-5d576c58-4cd1-43a5-b777-fc1f7d1f63db.PNG)

### Describe the solution you would like

Change the `role` attribute of the hidden input element from `text` to `combobox` so screen readers read out "combobox".

**Comment by melloware** [comment]
@Bukama I am slightly confused.  Right now that is `type="text"` and there is no `type="combobox"` its a type not a role?  So are you suggesting just adding `role="combobox"` to that hidden?

**Comment by Bukama** [comment]
> @Bukama I am slightly confused. Right now that is `type="text"` and there is no `type="combobox"` its a type not a role? So are you suggesting just adding `role="combobox"` to that hidden?

Sorry! Yes I meant to add `role` to it.

**Comment by melloware** [comment]
Got it!

**Comment by Bukama** [comment]
I know this issue was already closed at it seems do be done, but maybe my suggestion to fix this problem was not complete. When my team locally changed the role to combobox the screenreader seems happy, but the HTML-Validator shouts out a missing `aria-expanded` attribute - at least in disabled state.

@melloware Could you verifiy that  behavior with the upcoming PF 13 sources? Official showcase is (of course) still on PF 12

**Comment by melloware** [comment]
I guess the real question is not what the Checker says but what do your users report. Is the screen reader reading the  correct thing when disabled?

**Comment by christophs78** [comment]
IMO we should look what PrimeVue does. My disabled colleagues say PrimeVue works, but PrimeFaces SelectOneMenu does not work for them. (Jaws Screenreader)

**Comment by melloware** [comment]
Agreed we probably need to look at PrimeVue and emulate that if we can I know our SelectOne renders differently than the other Prime libs so it may not be as easy as we hope!

**Comment by christophs78** [comment]
@Bukama : Can you please have a look at https://www.primefaces.org/primevue/dropdown? Is this fully accessible? (PrimeTek claims it´s accessible and my disabled colleagues say it work´s for them using Jaws.)

**Comment by christophs78** [comment]
Which HTML-code is generated by PrimeVue-Dropdown? (as of 3.18.1)

![image](https://user-images.githubusercontent.com/10461942/200191553-3a27aa4f-9d51-47f6-8042-a4d7b70e1514.png)

![image](https://user-images.githubusercontent.com/10461942/200191686-2612f3ef-678f-4e7f-ae01-c04dabca3361.png)

**Comment by christophs78** [comment]
Code generated by PrimeFaces SelectOneMenu (as of 12.0.0)

![image](https://user-images.githubusercontent.com/10461942/200192639-5992ab0d-1cc2-413a-ba6c-5773208a89e6.png)

![image](https://user-images.githubusercontent.com/10461942/200192006-099f5806-3b17-42e0-af1f-fe1b94ea0aff.png)

**Comment by christophs78** [comment]
Cross-Check Angular Material - component "Select" (https://material.angular.io/components/select/overview, as of 14.2.6)

![image](https://user-images.githubusercontent.com/10461942/200192244-5e0bcda5-f46e-4737-b8b2-e9c992ced76b.png)

![image](https://user-images.githubusercontent.com/10461942/200192379-3b40c772-6cbd-4c9a-b65b-5cdf4cf9c1b8.png)

**Comment by christophs78** [comment]
After talking to my disabled colleague, thinking about this issue and comparing with PrimeVue und Angular Material i think we may try to get rid of the focus-input with text="text". The other two implementations work pretty fine without this. They assign role="combobox" to some DOM-element (eg span).
See https://github.com/primefaces/primevue/blob/master/src/components/dropdown/Dropdown.vue. (Input only for editable Dropdown, all other cases span via v-if="editable", v-else)

**Comment by christophs78** [comment]
Former work on this aka why is it currently implemented the way it is:

- focus-input was introduced with https://github.com/primefaces/primefaces/commit/4bdf60e7
  (was introduced to ensure clicking the label (referring via for-attribute to SelectOneMenu) set´s the focus to the SelectOneMenu)
- additional screereader-support was added with https://github.com/primefaces/primefaces/issues/598
- even more (too much) aria-attribues added with https://github.com/primefaces/primefaces/issues/4660
- and more aria-atributes added with https://github.com/primefaces/primefaces/issues/8473

**Comment by christophs78** [comment]
References (one more time):

- https://www.w3.org/WAI/ARIA/apg/example-index/combobox/combobox-select-only.html
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/combobox_role

**Comment by ZeroBrushV2** [comment]
Using the filter feature on SelectOneMenu it will generate a input in the dropdown where you can write down what to search.

The whole panel for the selectedOneMenu HTML will look like this:
```
<div id="j_idt319:optionSearch_panel"
    class="ui-selectonemenu-panel ui-widget ui-widget-content ui-corner-all ui-helper-hidden ui-shadow ui-input-overlay"
    style="
    display: block;
    width: 376.388px;
    left: 696.388px;
    top: 246.4px;
    transform-origin: center bottom;
    z-index: 1003;
  ">
    <div class="ui-selectonemenu-filter-container">
        <input class="ui-selectonemenu-filter ui-inputfield ui-inputtext ui-widget ui-state-default ui-corner-all"
            id="j_idt319:optionSearch_filter" name="j_idt319:optionSearch_filter" type="text" autocomplete="off"
            aria-autocomplete="list" aria-controls="j_idt319:optionSearch_table" aria-label="Filter" /><span
            class="ui-icon ui-icon-search"></span>
    </div>
    <div class="ui-selectonemenu-items-wrapper" style="max-height: 200px">
        <ul id="j_idt319:optionSearch_items"
            class="ui-selectonemenu-items ui-selectonemenu-list ui-widget-content ui-widget ui-corner-all ui-helper-reset"
            role="listbox" aria-activedescendant="j_idt319:optionSearch_0">
            <li class="ui-selectonemenu-item ui-selectonemenu-list-item ui-corner-all ui-state-highlight" tabindex="-1"
                role="option" data-label="Select One" aria-selected="true" id="j_idt319:optionSearch_0">
                Select One
            </li>
            <li class="ui-selectonemenu-item ui-selectonemenu-list-item ui-corner-all" tabindex="-1" role="option"
                data-label="Option1" aria-selected="false" id="j_idt319:optionSearch_1">
                Option1
            </li>
            <li class="ui-selectonemenu-item ui-selectonemenu-list-item ui-corner-all" tabindex="-1" role="option"
                data-label="Option2" aria-selected="false" id="j_idt319:optionSearch_2">
                Option2
            </li>
            <li class="ui-selectonemenu-item ui-selectonemenu-list-item ui-corner-all" tabindex="-1" role="option"
                data-label="Option3" aria-selected="false" id="j_idt319:optionSearch_3">
                Option3
            </li>
        </ul>
    </div>
    <div class="ui-selectonemenu-footer">
        <div id="j_idt319:j_idt413" role="separator"
            class="ui-divider ui-widget mt-0 ui-divider-horizontal ui-divider-solid ui-divider-left"></div>
        <span style="font-weight: bold">3 options</span>
    </div>
</div>
```

Let's focus on the problem, the input for the filter:
```
<input class="ui-selectonemenu-filter ui-inputfield ui-inputtext ui-widget ui-state-default ui-corner-all"
    id="j_idt319:optionSearch_filter" name="j_idt319:optionSearch_filter" type="text" autocomplete="off"
    aria-autocomplete="list" aria-controls="j_idt319:optionSearch_table" aria-label="Filter" />
```

The attribute `aria-controls` refers to an id that's not present in the page at all.

### Expected behavior

The aria-controls attribute should be like this:
`aria-controls=j_idt319:optionSearch_items`.
**NOTE: AS FOR MY DISCRETION I CHOOSE THE VALUE INSIDE THE ATTRIBUTE AS THE ONE THAT FITS THE WAI-ARIA STANDARD. AN OPEN DISCUSSION COULD VOTE THE APPROPRIATE ID TO INSERT IN THE ATTRIBUTE**

As per WAI-ARIA combobox standard [here](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/combobox_role):

> When a combobox's popup is displayed, ensure the [aria-controls](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls) attribute on the combobox element is set to the [id](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes#id) of the popup listbox, tree, grid, or dialog element. This is how the relationship between the element with the combobox role and the popup it controls is indicated. (Note: In older ARIA specs, this was aria-owns rather than aria-controls, so you may see aria-owns in older combobox implementations. The aria-owns in the code should be updated to aria-controls!))
